android/src/main/java/org/mozilla/experiments/nimbus/VariablesExamples.kt

```diff
@@ -0,0 +1,168 @@
+// This one is configuring feature1.
+fun configureFeature1() {
+    exampleJson = """
+        {
+            "buttonColor": "red",
+            "buttonText": "treatment_button_label"
+        }
+    """.trimIndent()
+
+    val config = nimbus.getFeatureVariables("example1") { EagerConfig1(it) }
+    val lazyConfig = nimbus.getFeatureVariables("example1") { LazyConfig1(it) }
+
+    val button = Button(context).apply {
+        text = config.buttonText
+        setBackgroundColor(config.buttonColor)
+    }
+
+    if (BuildConfig.DEBUG && config.buttonText != lazyConfig.buttonText) {
+        error("Assertion failed")
+    }
+
+    if (BuildConfig.DEBUG && config.buttonColor != lazyConfig.buttonColor) {
+        error("Assertion failed")
+    }
+}
+
+// This example does all its work in a secondary constructor, then discards to the `variables` object.
+// This has this has the benefit of being easy to use in a testing environment.
+// If we know that every key is going to be used, this is the version to use.
+// Because it has multiple constructors, this one (the data-class) will be better for re-use.
+data class EagerConfig1(
+    val buttonColor: Int,
+    val buttonText: String
+) {
+    constructor(variables: Variables) : this(
+        variables.getColor("buttonColor") ?: Color.BLUE,
+        variables.getText("buttonText") ?: "Okay then"
+    )
+}
+
+// This example does its work bit by bit, lazily grabbing the values out of JSON as it goes.
+// If we think that not every key is going to be used, then this will be the less CPU intensive.
+class LazyConfig1(
+    private val variables: Variables
+) {
+    val buttonColor: Int by lazy { variables.getColor("buttonColor") ?: Color.BLUE }
+    val buttonText: String by lazy { variables.getText("buttonText") ?: "Okay then" }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////
+// This one is configuring feature2. This will include a subset of variables:
+// for a button. In reality we'll not be doing anything this complex with something so simple.
+fun configureFeature2() {
+    exampleJson = """
+        {
+            "title": "Self Destruct",
+            "description": "Are you sure you wish to destroy the ship?",
+            "positiveButton": {
+                "color": "red",
+                "text": "Continue"
+            },
+            "negativeButton": {
+                "text": "No, cancel"
+            }
+        }
+    """.trimIndent()
+
+    val config = nimbus.getFeatureVariables("feature2") { SelfDestructDialogConfig(it) }
+
+    val dialogBuilder = AlertDialog.Builder(context).apply {
+        setTitle(config.title)
+        setMessage(config.description)
+        setIcon(config.icon)
+
+        setPositiveButton(config.positiveButton.text) { _, _ -> /* on click */ }
+        setNegativeButton(config.negativeButton.text) { dialog, _ ->
+            dialog.cancel()
+        }
+        create()
+    }
+
+    dialogBuilder.show()
```

Comment:
This is the application code. Everything else is to show the feature developers experience.

Reply:
To check my understanding - I interpret this as `configureFeature1` and `configureFeature2` containing the code that an application developer would write, and the other classes etc being code that we would (aspirationally) generate from a higher-level feature description or would (perhaps more likely in the short term) have the developer of the feature provide for use by the application code. Is that accurate?

Reply:
Yes, this is accurate.

I've split the code up into fake generated code and feature code that an application developer may write.

My aim here is to establish a reasonable half way house between app developers dealing with raw JSON like animals, and the full magic of a code generated future.